src/olympia/abuse/models.py

```diff
@@ -291,22 +293,41 @@ def resolve_job(self, *, log_entry):
         else:
             self.pending_rejections.clear()
         if cinder_decision.addon_id:
+            # We don't want to clear a NeedsHumanReview caused by a job that
+            # isn't resolved yet, but there is no link between NHR and jobs.
+            # So for each possible reason, we look if there are unresolved jobs
+            # and only clear NHR for that reason if there aren't any jobs left.
+            base_unresolved_jobs_qs = (
+                self.__class__.objects.for_addon(cinder_decision.addon)
+                .unresolved()
+                .resolvable_in_reviewer_tools()
+            )
             if was_escalated:
-                reasons = [NeedsHumanReview.REASONS.CINDER_ESCALATION]
-            else:
-                reasons = (
-                    [NeedsHumanReview.REASONS.ADDON_REVIEW_APPEAL]
-                    if self.is_appeal
-                    else []
+                has_unresolved_jobs_with_similar_reason = (
+                    base_unresolved_jobs_qs.filter(
+                        decision__action=DECISION_ACTIONS.AMO_ESCALATE_ADDON
+                    ).exists()
                 )
-                if self.abusereport_set.exists():
-                    reasons.append(NeedsHumanReview.REASONS.ABUSE_ADDON_VIOLATION)
-            NeedsHumanReview.objects.filter(
-                version__addon_id=cinder_decision.addon_id,
-                is_active=True,
-                reason__in=reasons,
-            ).update(is_active=False)
-            cinder_decision.addon.update_all_due_dates()
+                reason = NeedsHumanReview.REASONS.CINDER_ESCALATION
+            elif self.is_appeal:
+                has_unresolved_jobs_with_similar_reason = (
+                    base_unresolved_jobs_qs.filter(
+                        appealed_decisions__isnull=False
+                    ).exists()
+                )
+                reason = NeedsHumanReview.REASONS.ADDON_REVIEW_APPEAL
+            else:
+                # If the job we're resolving was not an appeal or escalation
+                # then all abuse reports are considered dealt with.
+                has_unresolved_jobs_with_similar_reason = None
+                reason = NeedsHumanReview.REASONS.ABUSE_ADDON_VIOLATION
+            if not has_unresolved_jobs_with_similar_reason:
```

Comment:
might be missing an edge case, but we could just not set `reason` to a non-`None` value unless the queryset exists, then we don't need this `has_unresolved_jobs_with_similar_reason` variable.

Reply:
Like this?

```diff
diff --git a/src/olympia/abuse/models.py b/src/olympia/abuse/models.py
index 0118d97338..f4da6a7cf7 100644
--- a/src/olympia/abuse/models.py
+++ b/src/olympia/abuse/models.py
@@ -302,26 +302,22 @@ class CinderJob(ModelBase):
                 .unresolved()
                 .resolvable_in_reviewer_tools()
             )
+            reason = None
             if was_escalated:
-                has_unresolved_jobs_with_similar_reason = (
-                    base_unresolved_jobs_qs.filter(
-                        decision__action=DECISION_ACTIONS.AMO_ESCALATE_ADDON
-                    ).exists()
-                )
-                reason = NeedsHumanReview.REASONS.CINDER_ESCALATION
+                if base_unresolved_jobs_qs.filter(
+                    decision__action=DECISION_ACTIONS.AMO_ESCALATE_ADDON
+                ).exists():
+                    reason = NeedsHumanReview.REASONS.CINDER_ESCALATION
             elif self.is_appeal:
-                has_unresolved_jobs_with_similar_reason = (
-                    base_unresolved_jobs_qs.filter(
-                        appealed_decisions__isnull=False
-                    ).exists()
-                )
-                reason = NeedsHumanReview.REASONS.ADDON_REVIEW_APPEAL
+                if base_unresolved_jobs_qs.filter(
+                    appealed_decisions__isnull=False
+                ).exists():
+                    reason = NeedsHumanReview.REASONS.ADDON_REVIEW_APPEAL
             else:
                 # If the job we're resolving was not an appeal or escalation
                 # then all abuse reports are considered dealt with.
-                has_unresolved_jobs_with_similar_reason = None
                 reason = NeedsHumanReview.REASONS.ABUSE_ADDON_VIOLATION
-            if not has_unresolved_jobs_with_similar_reason:
+            if reason:
                 NeedsHumanReview.objects.filter(
                     version__addon_id=cinder_decision.addon_id,
                     is_active=True,
```

I'm not sure this is better, it feels harder to read for me (though I agree the `has_unresolved_jobs_with_similar_reason = None` in my original version is not ideal)

Reply:
yeah, like that.  But I do see it's not a whole lot less code with all the line wrapping and wordy queries and constants.  It'd probably read better if we had queryset methods for `.escalated` and `is_appeal` but that's even more extra code... so 🤷 your choice on whatever you think is best.

Reply:
Ok, going to keep my original, slightly more verbose version for now. I suspect this is not the last time we'll need to refactor this code anyway...